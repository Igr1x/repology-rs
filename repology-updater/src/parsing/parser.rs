// SPDX-FileCopyrightText: Copyright 2025 Dmitry Marakasov <amdmi3@amdmi3.ru>
// SPDX-License-Identifier: GPL-3.0-or-later

use std::path::Path;

use crate::package::ParsedPackage;
use crate::parsing::error::PackageParsingError;
use crate::parsing::package_maker::PackageMaker;

pub trait PackageSink {
    fn push(&mut self, package_maker: PackageMaker) -> Result<(), PackageParsingError>;
}

pub trait PackageParser {
    fn parse(&self, path: &Path, sink: &mut dyn PackageSink) -> anyhow::Result<()>;
}

#[derive(Default)]
pub struct PackageAccumulator {
    pub packages: Vec<ParsedPackage>,
}

impl PackageSink for PackageAccumulator {
    fn push(&mut self, package_maker: PackageMaker) -> Result<(), PackageParsingError> {
        self.packages.push(package_maker.finalize()?);
        Ok(())
    }
}
