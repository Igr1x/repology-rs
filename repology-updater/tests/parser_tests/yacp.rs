// SPDX-FileCopyrightText: Copyright 2025 Dmitry Marakasov <amdmi3@amdmi3.ru>
// SPDX-License-Identifier: GPL-3.0-or-later

use std::path::Path;

use repology_updater::parsing::parser::PackageAccumulator;
use repology_updater::parsing::parser::PackageParser as _;
use repology_updater::parsing::parsers::yacp::YacpParser;

#[test]
fn test_yacp() {
    let mut packages = PackageAccumulator::default();
    YacpParser {}
        .parse(Path::new("tests/parser_tests/yacp/ok"), &mut packages)
        .unwrap();
    insta::assert_debug_snapshot!(packages.packages);
}
