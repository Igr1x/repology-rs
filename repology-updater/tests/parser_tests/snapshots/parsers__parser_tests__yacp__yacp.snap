---
source: repology-updater/tests/parser_tests/yacp.rs
expression: packages.packages
---
[
    Package {
        srcname: "ACE",
        binnames: [
            "ACE",
        ],
        trackname: "ACE",
        visiblename: "ACE",
        projectname_seed: "ACE",
        rawversion: "1.01",
        category: "Science",
        comment: "Accurate Correction of Errors",
        links: [
            Link {
                type: UpstreamHomepage,
                url: "https://github.com/Sheikhizadeh/ACE",
                fragment: None,
            },
        ],
        version: "1.01",
        flags: PackageFlags(
            0x0,
        ),
    },
]
